File: Boba.Core/Abelian.fs
namespace Boba.Core

module Abelian =

    open System
    open System.Diagnostics
    open Common
    open Fresh

    /// Represents a simple Abelian equation composed of constant values and variables which can each have a signed integer exponent.
    /// The implementation uses dictionaries as a form of signed multiset, where an element in the set can have more than one occurence
    /// (represented by a positive exponent) or even negative occurences (represented by a negative exponent). If an element has exactly zero
    /// occurences, it is removed from the dictionary for efficiency.
    [<DebuggerDisplay("{ToString()}")>]
    type Equation<'a, 'b> when 'a: comparison and 'b: comparison (variables: Map<'a, int>, constants: Map<'b, int>) =

        new(name: 'a) =
            Equation<'a, 'b>(Map.add name 1 Map.empty, Map.empty)

        new() =
            Equation<'a, 'b>(Map.empty, Map.empty)

        /// The set of variables in the unit equation, mapped to their exponents.
        member this.Variables = variables
        /// The set of constants in the unit equation, mapped to their exponents.
        member this.Constants = constants

        member this.IsIdentity () = this.Variables.IsEmpty && this.Constants.IsEmpty

        member this.IsConstant () = this.Variables.IsEmpty

        /// Negates each exponent in the equation.
        member this.Invert () =
            new Equation<'a, 'b>(mapValues (( ~- )) this.Variables, mapValues (( ~- )) this.Constants)

        /// Combine two Abelian unit equations. Values that appear in both equations have their exponents multiplied.
        member this.Add (other: Equation<'a, 'b>) =
            let mergeAdd (v1, v2) = v1 + v2
            let isExponentNonZero _ v = v <> 0
            let vars = mapUnion mergeAdd this.Variables other.Variables |> Map.filter isExponentNonZero
            let constants = mapUnion mergeAdd this.Constants other.Constants |> Map.filter isExponentNonZero
            new Equation<'a, 'b>(vars, constants)

        /// Removes the given equation from this Abelian unit equation. Equivalent to `this.Add(other.Invert())`.
        member this.Subtract (other: Equation<'a, 'b>) = other.Invert() |> this.Add;

        // Get the free variables of this equation.
        member this.Free () = mapKeys this.Variables
        
        member this.FractionString () =
            let expToStr exps =
                Map.map (fun k v -> if v = 1 then $"{k}" else $"{k}^{v}") exps |> Map.toList |> List.map snd
            let posVars = Map.filter (fun k v -> v >= 0) this.Variables |> expToStr
            let posCons = Map.filter (fun k v -> v >= 0) this.Constants |> expToStr
            let negVars = Map.filter (fun k v -> v < 0) this.Variables |> Map.map (fun k v -> -v) |> expToStr
            let negCons = Map.filter (fun k v -> v < 0) this.Constants |> Map.map (fun k v -> -v) |> expToStr
            let pos = String.concat "*" (List.append posVars posCons)
            let neg = String.concat "*" (List.append negVars negCons)
            if neg.Length > 0
            then pos + "/" + neg
            elif pos.Length <= 0
            then "1"
            else pos

        override this.GetHashCode() =
            hash (this.Variables, this.Constants)

        override this.Equals(b) =
            match b with
            | :? Equation<'a, 'b> as p -> (this.Variables, this.Constants) = (p.Variables, p.Constants)
            | _ -> false

        override this.ToString() =
            if this.IsIdentity()
            then "-"
            else
                let vars = Map.map (fun k v -> $"{k}^{v}") this.Variables |> Map.toList |> List.map snd
                let cons = Map.map (fun k v -> $"{k}^{v}") this.Constants |> Map.toList |> List.map snd
                String.concat "*" (List.append vars cons)
        
        interface IComparable<Equation<'a, 'b>> with
            member this.CompareTo(other: Equation<'a, 'b>) =
                let x = compare this.Variables other.Variables
                if x = 0 then compare this.Constants other.Constants else x
        
        interface IComparable with
            member this.CompareTo(other: obj) =
                match other with
                | :? Equation<'a, 'b> ->
                    let x = compare this.Variables (unbox<Equation<'a, 'b>> other).Variables
                    if x = 0 then compare this.Constants (unbox<Equation<'a, 'b>> other).Constants else x
                | _ -> invalidArg "other" "Must be of type Equation<'a, 'b>"

    let matchEqns (fresh: FreshVars) (eqn1 : Equation<string, 'b>) (eqn2 : Equation<string, 'b>) =
        let mgu (vars : List<string>) constVars constRigids (subst : Map<int, Linear.Equation>) =
            let toTerm freshVars vars consts =
                (new Equation<string, 'b>(List.zip freshVars vars |> Map.ofList, Map.empty))
                    .Add(new Equation<string, 'b>(List.zip constVars (List.take (List.length constVars) consts) |> Map.ofList, Map.empty))
                    .Add(new Equation<string, 'b>(Map.empty, List.zip constRigids (List.skip (List.length constVars) consts) |> Map.ofList))
            let toEquation freshVars acc (var, ind) =
                match Map.tryFind ind subst with
                | Some eqn -> Map.add var (toTerm freshVars eqn.Coefficients eqn.Constants) acc
                | None -> Map.add var (new Equation<string, 'b>(freshVars.[ind])) acc
            let freshies =
                if Map.isEmpty subst
                then []
                else
                    Map.toList subst
                    |> List.head
                    |> snd
                    |> (fun (a : Linear.Equation) -> a.Coefficients)
                    |> List.length
                    |> fresh.FreshN "a"
            List.fold (toEquation freshies) Map.empty (List.mapi (fun i b -> (b, i)) vars)

        if eqn1.Variables.IsEmpty && eqn2.Variables.IsEmpty
        then
            if eqn1.Constants = eqn2.Constants
            then Some Map.empty
            else None
        elif eqn1.Variables.IsEmpty
        then None
        else
            let bases map = Map.toList map |> List.map fst
            let exponents map = Map.toList map |> List.map snd
            // put all constants on the 'constant' side of the equation, so that the matching side only has variables
            let right = eqn2.Subtract(new Equation<string, 'b>(Map.empty, eqn1.Constants))
            Linear.solve { Coefficients = exponents eqn1.Variables; Constants = List.append (exponents right.Variables) (exponents right.Constants) }
            |> Option.map (mgu (bases eqn1.Variables) (bases right.Variables) (bases right.Constants))

    let unify (fresh: FreshVars) (eqn1 : Equation<string, 'b>) (eqn2 : Equation<string, 'b>) =
        matchEqns fresh (eqn1.Add(eqn2.Invert())) (new Equation<string, 'b>())

File: compiler/abelian/abelian.go
package abelian

import (
	"github.com/glossopoeia/boba/compiler/linear"
	"github.com/glossopoeia/boba/compiler/util"
	"golang.org/x/exp/constraints"
	"golang.org/x/exp/maps"
	"golang.org/x/exp/slices"

	"github.com/rjNemo/underscore"
)

// Represents an Abelian equation composed of constant values and variables,
// each of which can have a signed integer exponent. The implementation uses
// dictionaries as a form of signed multiset, where an element in the set can
// have more than one occurence (represented by a positive exponent) or even
// negative occurences (represented by a negative exponent). If an element has
// exactly zero occurences, it is removed from the dictionary for efficiency.
type Equation[K constraints.Ordered, V constraints.Ordered] struct {
	// The variables of the equation and their associated exponents.
	Variables map[K]int
	// The constants of the equation and their associated exponents.
	Constants map[V]int
}

// A substitution on abelian equations, where K is the type of the variables
// in the equation and V the type of the constants.
type Substitution[K constraints.Ordered, V constraints.Ordered] map[K]Equation[K, V]

// Create the identity Abelian equation, with no variables and no constants.
func AbelianIdentity[K constraints.Ordered, V constraints.Ordered]() Equation[K, V] {
	return Equation[K, V]{make(map[K]int), make(map[V]int)}
}

// Create a single variable Abelian equation.
func AbelianVariable[K constraints.Ordered, V constraints.Ordered](varName K) Equation[K, V] {
	return Equation[K, V]{map[K]int{varName: 1}, map[V]int{}}
}

// True if the equation has no variables and no constants.
func (eqn Equation[K, V]) IsIdentity() bool {
	return len(eqn.Variables) == 0 && len(eqn.Constants) == 0
}

// True if the equation has no variables.
func (eqn Equation[K, V]) IsConstant() bool {
	return len(eqn.Variables) == 0
}

// Get the variables of the equation without their associated exponents.
func (eqn Equation[K, V]) Free() []K {
	return maps.Keys(eqn.Variables)
}

// Get the exponent of the given variable name within this equation.
// Returns 0 if the variable is not present.
func (eqn Equation[K, V]) ExponentOf(v K) int {
	if e, ok := eqn.Variables[v]; ok {
		return e
	}
	return 0
}

// Negate all the exponents in the equation.
func (eqn Equation[K, V]) Invert() Equation[K, V] {
	inverted := Equation[K, V]{maps.Clone(eqn.Variables), maps.Clone(eqn.Constants)}
	for k, v := range inverted.Variables {
		inverted.Variables[k] = -v
	}
	for k, v := range inverted.Constants {
		inverted.Constants[k] = -v
	}
	return inverted
}

// Combine two Abelian equations via summing. Values that appear in both equations
// have their exponents multiplied.
func (eqn Equation[K, V]) Add(other Equation[K, V]) Equation[K, V] {
	mergeAdd := func(l, r int) int { return l + r }
	expNotZero := func(v int) bool { return v != 0 }
	vars := util.MergeMaps(eqn.Variables, other.Variables, mergeAdd)
	consts := util.MergeMaps(eqn.Constants, other.Constants, mergeAdd)
	vars = util.MapFilterValue(vars, expNotZero)
	consts = util.MapFilterValue(consts, expNotZero)
	return Equation[K, V]{vars, consts}
}

// Removes the given equation from this Abelian unit equation. Equivalent to `eqn.Add(other.Invert())`.
func (eqn Equation[K, V]) Subtract(other Equation[K, V]) Equation[K, V] {
	return eqn.Add(other.Invert())
}

// Generate a substitution that, when applied to eqn, makes it equal to other with
// respect to the equational rules for Abelian systems. The approach employed is
// that of solving linear diophantine equations. If there is no way to match the
// equations, the resulting substitution is nil.
func (eqn Equation[K, V]) Match(fresh util.Fresh[K], other Equation[K, V]) Substitution[K, V] {

	if eqn.IsConstant() && other.IsConstant() {
		if maps.Equal(eqn.Constants, other.Constants) {
			return Substitution[K, V]{}
		}
		return nil
	}
	if eqn.IsConstant() {
		return nil
	}

	// put all constants on the 'constant' side of the equation, so that the matching side only has variables
	right := other.Subtract(Equation[K, V]{map[K]int{}, eqn.Constants})

	// convert a few of the variable and constants maps into slices,
	// for multiple deterministic indexed traversals
	ordEqnVars := maps.Keys(eqn.Variables)
	ordRightVars := maps.Keys(right.Variables)
	ordRightConsts := maps.Keys(right.Constants)

	slices.Sort(ordEqnVars)
	slices.Sort(ordRightVars)
	slices.Sort(ordRightConsts)

	ordEqnExps := underscore.Map(ordEqnVars, func(k K) int { return eqn.Variables[k] })
	ordRightVarExps := underscore.Map(ordRightVars, func(k K) int { return right.Variables[k] })
	ordRightConstExps := underscore.Map(ordRightConsts, func(k V) int { return right.Constants[k] })

	// convert into a linear diophantine equation for solving
	linear := linear.Equation{
		Coefficients: ordEqnExps,
		Constants:    append(ordRightVarExps, ordRightConstExps...),
	}

	// if a solution exists, convert the linear equation substitution into an Abelian substitution
	if solution := linear.Solution(); solution != nil {
		// linear solution can have more variables that original equation
		varCount := len(maps.Values(solution)[0].Coefficients)
		fresh := fresh.NextN(varCount)

		// convert the linear substitution variable indexes into named variables
		varSubst := Substitution[K, V]{}
		for i, k := range ordEqnVars {
			// convert the solution equation into an Abelian equation
			if sub, ok := solution[i]; ok {
				// convert flex variables
				flexVarExps := map[K]int{}
				for fi, fv := range fresh {
					flexVarExps[fv] = sub.Coefficients[fi]
				}
				// convert constant variables
				constVarExps := map[K]int{}
				for ci, cv := range ordRightVars {
					constVarExps[cv] = sub.Constants[ci]
				}
				// convert constants
				constExps := map[V]int{}
				for ci, cc := range ordRightConsts {
					constExps[cc] = sub.Constants[ci+len(right.Variables)]
				}

				//combine converted subequations into a single equation
				combined := Equation[K, V]{flexVarExps, map[V]int{}}.Add(
					Equation[K, V]{constVarExps, map[V]int{}},
				).Add(
					Equation[K, V]{map[K]int{}, constExps},
				)
				varSubst[k] = combined
			} else {
				varSubst[k] = AbelianVariable[K, V](fresh[i])
			}
		}
		return varSubst
	}
	return nil
}

// Generate a most general unifier that, when applied to both eqn and other, makes the resulting
// Abelian equations equal with respect to the equational rules for Abelian systems.
// The approach employed is that of solving linear diophantine equations. If there is no way
// to unify the equations, the resulting substitution is nil.
func (eqn Equation[K, V]) Unify(fresh util.Fresh[K], other Equation[K, V]) Substitution[K, V] {
	return eqn.Add(other.Invert()).Match(fresh, AbelianIdentity[K, V]())
}
